### ISSUE TITLE
Better Fighters Through Squadrons

### ISSUE THREAD

**Issue by QuantumTux** [endless-sky/endless-sky#9671]
### Problem Description

Fighters are fragile and not terribly effective - they are basically missiles that avoid asteroids and will **_probably_** get to fire before they die.

Also, ss [Zitchas said in #9429](https://github.com/endless-sky/endless-sky/issues/9429#issuecomment-1752303435), "....we don't have carriers, we have CQC battleships that just so happen to carry a few fighters to support them."

This proposal puts forth ideas that may address these shortcomings; I hope that they are possible to implement **without** major impact to the AI.

### Related Issue Links

- [Issue #6523](https://github.com/endless-sky/endless-sky/issues/6523) - Interdiction and Combat Instancing for slower, more tactical combat
- [Issue #7539](https://github.com/endless-sky/endless-sky/issues/7539) - Just **Better AI** in the most general sense
- [Issue #7552](https://github.com/endless-sky/endless-sky/issues/7552) - Mechanics/Bugfix: better AI distancing when using artillery or explosives at short range
- [Issue #8033](https://github.com/endless-sky/endless-sky/issues/8033) - A way to make improvements on AI safer
- [Issue #9272](https://github.com/endless-sky/endless-sky/issues/9272) - Auto-Deploy & Recall for Carried Units
- [Issue #9429](https://github.com/endless-sky/endless-sky/issues/9429) - Carriers rarely get a chance to repair their fighters while in combat
- [Issue #9441](https://github.com/endless-sky/endless-sky/issues/9441) - feat(content): Add new ship variants optimised for long-range carrier tactics

### Desired Solution

In order to make Fighters "smarter", we can provide **Roles** for Fighters; a **Role** pre-defines their behavior when deployed. This pairs well with [Issue #9272](https://github.com/endless-sky/endless-sky/issues/9272). Also part of this proposal is the concept of **Squadron Leaders**, and pre-defined **Squadron formations** (patterns of deployment). To start laying this out, I define five different **Roles**:

1. **Defensive Screen**: Fighters station-keep[^1] with launching ship, and self-destribute around it using a pre-defined pattern (Formation). Fighters operating in this Role
- Only attack hostile Fighters (not ships!) that approach to within weapons range of a Fighter
- Will **not** give chase past a relatively small radius from the launching ship (less than the effective range of most anti-missile Outfits - maybe 250 to 300)
- If the Fighters are themselves equipped with anti-missile Outfits, they will preferentially target missiles headed for their launching ship (and any headed for the Fighters, too)
2. **Offensive Screen**: Fighters stay close (station-keep) with launching ship similar to **Defensive Screen**. Fighters operating in this Role:
- Will actively attack any ship (or Fighter) that hits their launching ship with a direct-fire weapon (but not a missile)
- Will chase a target out to a significant radius from their launching ship, but will not pursue if they reach that limit and their range-to-target is not decreasing. Suggested maximum radius would be between 1000 and 2000 (but Fighters will continue pursuit if they reach that point and their range to their target has been **decreasing** over the past few seconds).
- If the Fighters are themselves equipped with anti-missile Outfits, they will target preferentially missiles headed for their launching ship (and any headed for the Fighters, too)
3. **Attack Lead**: Fighters station-keep with the launching ship, and organize into a "flying wedge" formation, leading the attack
- They lead the attack by flying in front of their launching ship, but they do not allow the distance between the Fighters and their launching ship to exceed about half of the ship's maximum range for direct-fire weapons[^2]
- Their priority target is whatever their launching ship is targeting (but they attempt to maintain formation in front of their launching ship)
- Secondary target is any enemy Fighters that approach (but they will **NOT** give chase)
- Anti-missile Outfits employed as _per_ the **Defensive Screen**.
4. **Hunter Squadron**: Fighters self-organize into a squadron and fly in formation as much as possible. When operating in this Role, the Fighters:
- Actively attack targets as a group. Preferred targets, in **decreasing** order, are hostile Fighters, followed by hostile ships. Ships are attacked by size and distance - smallest and closest first. It may be possible to use a math formula to help the AI select; perhaps multiply the size of each hostile ship by some number (10? 20?), then add the range - they attack the lowest number first.
- When flying this mission Fighters do not attempt to stay within a certain range of their launching ship.
5. **Combat Air Patrol**: Fighters self-distribute into a squadron formation similar to "Hunter Squadron" above
- Will not station-keep with their launching ship but also will not stray beyond some multiple (2x? 3x? 5x?) of the maximum range of direct-fire weapons on the launching ship
- Will attack hostile units that come within weapons range of the Fighters
- Any hostile unit that hits the launching ship will become a target if the Fighters are not currently engaging another hostile target

The behavior of the Fighters is further refined by the **Aggression** setting of the Squadron:

1. **Offensive**: Attack until target is disabled/destroyed
- Ignore shield damage to self, but break off and return to carrier if damage to Hull exceeds 49%
- This setting is **forced** for Fighters in the following Roles: **Attack Lead**, **Hunter Squadron**
- This setting may be selected for Fighters in the following Roles: **Combat Air Patrol**
2. **Defensive**: Attack until target is destroyed/disabled; or, the target leaves effective weapon radius
- Break off if the Fighter shield is reduced to 0; return to carrier if damage to Hull exceeds 24%
- This setting is **forced** for Fighters in the following Roles: **Offensive Screen**
- This setting may be selected for Fighters in the following roles: **Combat Air Patrol**, **Defensive Screen**
3. **Passive**: Defend with anti-missile Outfits; do not attack any hostile ships, Drones or Fighters (essentially, only use Anti-Missile outfits)
- Return to carrier if any Hull damage is taken
- This setting may be selected for Fighters in the following Roles: **Defensive Screen**

[^1]: **Station-keeping** is defined as the Fighters staying very close to the launching ship, say about 50 or so. They would only leave that radius if demanded by the Role
[^2]: There would need to be a mechanism for figuring the maximum distance allowed when the launching ship is a missile boat and has no direct-fire weapons

### Alternative Approaches

There may be other Roles to define; this idea gives a format for doing so.

When considering alternatives, this proposal establishes a short list of rules for the AI to follow when maneuvering Fighters
- Max allowed distance from launching ship
- Rules of engagement/criteria for initiating and/or breaking off an attack
- Target selection and priority
- Self-preservation rules

This approach may (or may not) be helpful for Drones (but that would be outside the scope of this proposal)

### Additional Context

A **Squadron** is a group of up to six Fighters launched from the same ship. When a ship can carry more than six Fighters, then the number of Fighters is split as evenly as possible, into multiple Squadrons, with no more than six Fighters _per_ Squadron. When splitting the Fighters into multiple Squadrons, if there are multiple types of Fighters aboard the ship, the first criteria is to group the same Fighters together; then, one Squadron can have multiple types if that is needed to balance the numbers

To help the AI, a single Fighter launched from a ship is selected as the **Squadron Lead**; for all the Roles. This means the AI does not need to select actions for each individual Fighter - rather, the Squadron Leader is the focus of the AI decision-making process, and all the other Fighters in the Squadron follow the Squadron Leader. The **Squadron Lead Fighter** selects targets for all Fighters in the Squadron, and is the Fighter checked to see if the Squadron attacks.

The Squadron Lead is "elected" when the Fighters are first deployed; the leader is randomly selected from the Fighters, unless the launching ship is carrying multiple Fighter types, in which case the lead is randomly selected from the type with the highest Hull value. Additionally, if:
- the Squadron Lead has to land, or is disabled/destroyed, then the remaining Fighters "elect" a Squadron Lead using the same rules
- a Fighter that was Squadron Lead is redeployed/repaired, this does **not** trigger a new election

Also to help the AI, we can define **Squadron Formations** for each Role. These are proposed deployment patterns for various numbers of Fighters in the Squadron, using the following tokens:
- **[S]** indicates the **Squadron Leader**
- **[F]** indicates the other Fighters in the **Squadron**
- **[LS]** indicates the launching ship

For the **Defensive Screen** and **Offensive Screen** Roles:
```
  [S]     [S][LS][F]                [S]  [F]       [S]      [S]  [F]
  [LS]                   [S]          [LS]     [F][LS][F]  [F][LS][F]
                      [F][LS][F]    [F]  [F]    [F]  [F]    [F]  [F]
```
For the **Attack Lead** Role:
```
  [S]   [S]  [F]      [S]       [S][F]       [S]          [S][F]
  [LS]    [LS]     [F]   [F]  [F]   [F]    [F][F]       [F]    [F]
                      [LS]      [LS]     [F]    [F]   [F]        [F]
                                            [LS]           [LS]
```
For the **Hunter Squadron** and **Combat Air Patrol** Roles:
```
  [S]     [S][F]      [S]       [S][F]        [S]            [S][F]
                    [F] [F]   [F]    [F]    [F] [F]     [F][F]    [F][F]
                                          [F]     [F]
```

**Comment by yafflehk** [comment]
Make all turrets above a certain size/weight/DPS ignore or be to slow to track fighters, make fighters vulnerable to anti-missile defenses.

**Comment by Zitchas** [comment]
@petervdmeer did a lot of work on formations in https://github.com/endless-sky/endless-sky/pull/4471. With taking a look for some ideas on how to set up formations. It was an awesome PR, but it got killed off. 

Anyway, you have a lot of good ideas, would make fighters a lot more useful. I look forward to seeing what you do.

**Comment by QuantumTux** [comment]
@Zitchas @petervdmeer I agree #4471 is excellent foundational work towards the shared goal. I don't have the same level of knowledge with the ES code internals, but it seems like those changes would handle a lot of basic mechanics of making formation-flying happen.

I hope my suggestions _vis-a-vis_ centralizing the AI decision-making (Squadrons and Squadon Leaders, Roles and Aggression) will help make the rest of the implementation workable.
